core/src/main/java/org/renjin/sexp/Environment.java

```diff
@@ -216,27 +216,78 @@ public StringVector getNames() {
   public boolean bindingIsLocked(Symbol symbol) {
     return lockedBindings != null && lockedBindings.contains(symbol);
   }
-  
-  public void setVariable(String name, SEXP value) {
+
+  public void setVariableOnlyIfThereAreNoActiveBindings(Context context, Symbol symbol, SEXP value) {
+    assert ( activeBindings == null );
+
+    if(value == Symbol.UNBOUND_VALUE) {
+      throw new EvalException("Unbound: " + symbol);
+    }
+
+    if(bindingIsLocked(symbol)) {
+      throw new EvalException("cannot change value of locked binding for '%s'", symbol.getPrintName());
+    } else if(locked && frame.getVariable(symbol) == Symbol.UNBOUND_VALUE) {
+      throw new EvalException("cannot add bindings to a locked environment");
+    }
+    frame.setVariable(symbol, value);
+    modCount++;
+  }
+
+  public void setVariableOnlyIfThereAreNoActiveBindings(Context context, String name, SEXP value) {
+    assert ( context != null );
     if(StringVector.isNA(name)) {
       name = "NA";
     }
-    setVariable(Symbol.get(name), value);
+    setVariableOnlyIfThereAreNoActiveBindings(context, Symbol.get(name), value);
   }
 
-  public void setVariable(Symbol symbol, SEXP value) {
-    
+  public void setVariable(Context context, String name, SEXP value) {
+    assert ( context != null );
+    if(StringVector.isNA(name)) {
+      name = "NA";
+    }
+    setVariable(context, Symbol.get(name), value);
+  }
+
+  public void setVariable(Context context, Symbol symbol, SEXP value) {

```

Comment:
Add documentation that describes handling of active bindings, warns of side effects.